src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -156,39 +157,55 @@ module IncrementalBuildSyntaxTree =
                 fileParsed.Trigger fileName
 
                 input, sourceRange, fileName, diagnosticsLogger.GetDiagnostics()
+            )
 
+        let parse (parseTask: Threading.Tasks.Task<_>) =
+            try
+                parseTask.Result
             with exn ->
                 let msg = sprintf "unexpected failure in SyntaxTree.parse\nerror = %s" (exn.ToString())
                 System.Diagnostics.Debug.Assert(false, msg)
                 failwith msg
 
-        let parseOrSkip sigNameOpt =
+        let getValue source = lock source <| fun () -> cache.GetValue(source, getParseTask)
```

Comment:
Just a sanity check - is shadowed `source` fine to lock on in here.

Reply:
Should be fine, source is cache key and getParseTask does not wait.